internal/driver/common/canvas.go

```diff
@@ -288,7 +292,34 @@ func (c *Canvas) Painter() gl.Painter {
 
 // Refresh refreshes a canvas object.
 func (c *Canvas) Refresh(obj fyne.CanvasObject) {
-	c.refreshQueue.In(obj)
+	var objCopy copy.CopyCanvasObject
+	switch co := obj.(type) {
+	case *canvas.Circle:
+		objCopy = copy.NewCopyCircle(co)
+	case *fyne.Container:
+		objCopy = copy.NewCopyContainer(co)
+	case *canvas.Image:
+		objCopy = copy.NewCopyImage(co)
+	case *canvas.Line:
+		objCopy = copy.NewCopyLine(co)
+	case *canvas.LinearGradient:
+		objCopy = copy.NewCopyLinearGradient(co)
+	case *canvas.RadialGradient:
+		objCopy = copy.NewCopyRadialGradient(co)
+	case *canvas.Raster:
+		objCopy = copy.NewCopyRaster(co)
+	case *canvas.Rectangle:
+		objCopy = copy.NewCopyRectangle(co)
+	case *canvas.Text:
+		objCopy = copy.NewCopyText(co)
+	}
```

Comment:
I am skeptical that this could be done efficiently. In order to obtain the same performance. We would desire a fast path where no copy is involved. When it becomes necessary, we fall into a slow path and then do a copy. Copy is also not memory friendly and cumbersome on circular structures.

Reply:
Of all the type, Container and Widget are the most costly one to copy (due to their children). The others are really basic simple structure which copying is really not a problem, especially if we manage to avoid memory allocation.

The problem with Container and Widget could be improved upon if we have preallocated state. In this case, you only need to copy the children if themselves have changed compared to the last time a state was used. This would also help solving the SetContent call, as it could then be easier to commit on render the tree that is now valid.

I do not think that without multiple copy, we can actually solve logically this problem. There are different state that are needed by different part of the code with different value.
- A place for the user of the API to store the value until it call Refresh
- A place where the renderer read value when they are Refreshed
- A place where the input thread read value to figure out what is on screen to know where to dispatch events

Adding locks, channel or queue would not solve the logical need for each of this step to have potentially different value only copy allow for that possibility.

(The fourth state is to avoid a big lock between user api call to Refresh and the renderer)

Reply:
I think presumably we are getting too deeper here regarding implementation detail. We need first decide whether the initial assumption of this approach is really acceptable or not, which is: we ask all users to be aware of and guarantee their rendering-related code to run on the rendering thread.

> I do not think that without multiple copy, we can actually solve logically this problem.

Shared memory pattern has been researched for decades, there are many existing practices. Even with channels, it is underlying lock-free shared memory with few exceptions that semantically copies tiny values.

Reply:
I think it is actually this detailed discussion that shows the most that this is not a race condition problem first, but a logical one.

The value that are set on the object by the user, should not be assumed by the input layer to have reach the screen. Even if we had lock everywhere we can think of, it doesn't make sense. Only after this value have been rendered, does it makes sense for the input layer to use them.

The same apply for the rendering logic, it should not be using the same value as the input layer as it would mean it has already been rendered. It should not also use the value before a call to Refresh as it would mean that the user as not yet set his will to use this value on the next rendering operation.

With that in mind, it is logical that we have multiple tree in memory at all time, because they are different state and that the problematic is to synchronize them. If we agree on this, we can go back on the discussion on what the synchronization mechanism should be.

Reply:
I can hardly follow what you are trying to say. I am sorry :( It would be really nice if you could name a concrete example to clarify what is the problem here and what we are trying to solve.

The previous comments on my side were about the motivation of "we ask all users to be aware of and guarantee their rendering-related code to run on the rendering thread" is too strong.

Reply:
> I can hardly follow what you are trying to say. I am sorry :( It would be really nice if you could name a concrete example to clarify what is the problem here and what we are trying to solve.

If you modify any object, it is instantly reflected in the tree that the input layer and the rendering layer use. There is no need for calling Refresh for this change to be propagated. It should be possible for example to move/resize an object never have it go through rendering, but manage to intercept a mouse click. Their is no synchronization logic between any of the layer.
  The go race detection tool does only detect the fact we do not have that synchronization, not what is the underlying problem (not the lack of lock, but the lack of logical synchronization between steps).

> The previous comments on my side were about the motivation of "we ask all users to be aware of and guarantee their rendering-related code to run on the rendering thread" is too strong.

I did not understand this sentence, which is why I go back to try to explain the synchronization problem that lead to this race condition first. Currently, there is no way for the user to do any synchronization, so this sentence doesn't apply. In the PR here, I do a full copy of all the CanvasObject at Refresh time to be used by the rendering thread. This means the user does not need to be aware of the rendering thread, he must still be aware that his call to Refresh have to be synchronized between different goroutine. So I do not understand why you think the users have to be aware of the rendering thread at all.